src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorDataSourceRetriever.java

```diff
@@ -0,0 +1,221 @@
+        // Make sure the user specified at least one data source to download:
+        if ((!getSomaticDataSources) && (!getGermlineDataSources)) {
+            throw new UserException("Must select either somatic or germline datasources.");
+        }
+
+        if ( overwriteOutputFile ) {
+            logger.info("Overwrite ENABLED.  Will overwrite existing data sources download.");
+        }
+    }
+
+    //==================================================================================================================
+    // Static Methods:
+
+    //==================================================================================================================
+    // Instance Methods:
+
+    private NioFileCopier downloadDatasources(final Path dataSourcesPath) {
+
+        // Get the data sources file:
+        final Path outputDestination = getDataSourceLocalPath(dataSourcesPath);
+
+        // Set up and initiate our download:
+        final NioFileCopier xerox = NioFileCopier.create(dataSourcesPath, outputDestination, overwriteOutputFile);
+        xerox.initiateCopy();
+
+        return xerox;
+    }
+
+    private void validateIntegrity(final NioFileCopier fileCopier, final Path remoteSha256Path) {
+
+        // Get the SHA 256 file:
+        final Path localSha256Path = getSha256File(remoteSha256Path);
+
+        // Read the sha256sum into memory:
+        final String expectedSha256Sum = readSha256SumFromFile(localSha256Path);
+
+        // Calculate the sha256sum of the data sources file:
+        final boolean downloadedDataSourcesAreValid = fileCopier.validateIntegrity(expectedSha256Sum, DigestUtils::sha256Hex);
+
+        // verify the hashes are the same:
+        if ( !downloadedDataSourcesAreValid ) {
+            throw new GATKException("ERROR: downloaded data sources are corrupt!  Unexpected checksum: " + fileCopier.getLatestChecksum() + " != " + expectedSha256Sum);
+        }
+        else {
+            logger.info("Data sources are valid.");
+        }
+    }
+
+    private Path getSha256File(final Path remoteSha256Path) {
+        final Path localSha256Path = getDataSourceLocalPath(remoteSha256Path);
+        logger.info("Retrieving expected checksum file...");
+        NioFileCopier.create(remoteSha256Path, localSha256Path, overwriteOutputFile).setSilentCopy(true).initiateCopy();
+        logger.info("File transfer complete!");
+
+        return localSha256Path;
+    }
+
+    private String readSha256SumFromFile(final Path localSha256Path) {
+        final String expectedSha256Sum;
+        try {
+            logger.info("Collecting expected checksum...");
+            expectedSha256Sum = FileUtils.readFileToString(localSha256Path.toFile(), Charset.defaultCharset());
+            logger.info("Collection complete!");
+        }
+        catch ( final IOException ex ) {
+            throw new GATKException("Could not read in sha256sum from file: " + localSha256Path.toUri().toString(), ex);
```

Comment:
This is probably a `UserException` as well. Again, we generally blame the user for I/O issues out of our control.

Reply:
Fixed